<a href="https://colab.research.google.com/github/jenyadev3/TopResponders/blob/main/vkr_project_target_company.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Задача** - построить портрет клиента, склонного к позитивному реагированию на реклама

**Шаг 1. Загрузка данных**


In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv("/content/dataset_target_company.csv")

*Просмотр данных*

In [45]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15223 entries, 0 to 15222
Data columns (total 37 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  15223 non-null  int64  
 1   target              15223 non-null  int64  
 2   age                 15223 non-null  int64  
 3   is_employed         15223 non-null  int64  
 4   is_pensioner        15223 non-null  int64  
 5   gender              15223 non-null  int64  
 6   childs              15223 non-null  int64  
 7   dependants          15223 non-null  int64  
 8   education           15223 non-null  object 
 9   marital_status      15223 non-null  object 
 10  industry            13856 non-null  object 
 11  job_title           13856 non-null  object 
 12  org_type_state      13856 non-null  object 
 13  org_type_capital    13858 non-null  object 
 14  job_field           13856 non-null  object 
 15  family_income       15223 non-null  object 
 16  pers

In [43]:
df.describe()

id   target      age  is_employed  is_pensioner   gender  \
count    15223.00 15223.00 15223.00     15223.00      15223.00 15223.00   
mean  65401830.70     0.12    40.41         0.91          0.13     0.65   
std    4568181.14     0.32    11.60         0.29          0.34     0.48   
min   59910150.00     0.00    21.00         0.00          0.00     0.00   
25%   61920915.00     0.00    30.00         1.00          0.00     0.00   
50%   64370996.00     0.00    39.00         1.00          0.00     1.00   
75%   67088017.50     0.00    50.00         1.00          0.00     1.00   
max   75292242.00     1.00    67.00         1.00          1.00     1.00   

        childs  dependants  personal_income  has_real_state  ...  has_cottage  \
count 15223.00    15223.00         15223.00        15223.00  ...     15223.00   
mean      1.10        0.65         13853.84            0.31  ...         0.01   
std       1.00        0.81          9015.47            0.46  ...         0.10   
min       0.00        0.00            24.00            0.00  ...         0.00   
25%       0.00        0.00          8000.00            0.00  ...         0.00   
50%       1.00        0.00         12000.00            0.00  ...         0.00   
75%       2.00        1.00         17000.00            1.00  ...         0.00   
max      10.00        7.00        250000.00            1.00  ...         1.00   

       has_garage  has_land    credit     term  residence_duration  \
count    15223.00  15223.00  15223.00 15223.00            15223.00   
mean         0.02      0.04  14667.96     8.10             3039.34   
std          0.14      0.20  12147.87     4.09           262455.44   
min          0.00      0.00   2000.00     3.00              -26.00   
25%          0.00      0.00   6500.00     6.00               41.00   
50%          0.00      0.00  11550.00     6.00              108.00   
75%          0.00      0.00  19170.00    10.00              204.00   
max          1.00      1.00 119700.00    36.00         28101997.00   

       work_experience  number_of_payments  overdue_payments  max_of_amount  
count         13855.00            15223.00          15223.00       15223.00  
mean            292.21                7.88              0.20         243.02  
std           24364.83                7.96              0.69         836.10  
min               1.00                1.00              0.00           0.00  
25%              24.00                4.00              0.00           0.00  
50%              48.00                5.00              0.00           0.00  
75%             110.00                8.00              0.00           0.00  
max         2867959.00              114.00             13.00       28012.20  

[8 rows x 23 columns]

In [ ]:
df.head()

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
0      59910150       0   49                  1                  0       1   
1      59910230       0   32                  1                  0       1   
2      59910525       0   52                  1                  0       1   
3      59910803       0   39                  1                  0       1   
4      59911781       0   30                  1                  0       0   

   CHILD_TOTAL  DEPENDANTS            EDUCATION  MARITAL_STATUS  ...  \
0            2           1  Среднее специальное  Состою в браке  ...   
1            3           3              Среднее  Состою в браке  ...   
2            4           0     Неполное среднее  Состою в браке  ...   
3            1           1               Высшее  Состою в браке  ...   
4            0           0              Среднее  Состою в браке  ...   

  GAR_PRESENCE_FL LAND_PRESENCE_FL    CREDIT TERM FACT_LIVING_TERM WORK_TIME  \
0               0                0   8000.00    6              220     18.00   
1               0                0  21650.00    6              137     97.00   
2               0                0  33126.00   12              251     84.00   
3               0                0   8491.82    6               36    168.00   
4               0                0  21990.00   12               83    101.00   

   LOAN_NUM_PAYM LOAN_DLQ_NUM LOAN_MAX_DLQ_AMT          DATE_CREDIT  
0              6            2          1580.00  2017-10-01 15:09:09  
1              6            1          4020.00  2020-10-09 18:58:35  
2             11            0             0.00  2015-05-05 11:10:07  
3              6            3          1590.00  2016-06-19 10:05:49  
4             16            2          2230.00  2016-03-13 14:22:06  

[5 rows x 37 columns]

In [ ]:
df.tail()

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
15218      75291424       0   25                  1                  0   
15219      75291888       0   51                  1                  0   
15220      75291923       0   22                  1                  0   
15221      75291960       0   60                  0                  1   
15222      75292242       0   37                  1                  0   

       GENDER  CHILD_TOTAL  DEPENDANTS            EDUCATION  \
15218       0            1           1              Среднее   
15219       1            2           0  Среднее специальное   
15220       1            1           1              Среднее   
15221       1            2           0              Среднее   
15222       0            2           0              Среднее   

           MARITAL_STATUS  ... GAR_PRESENCE_FL LAND_PRESENCE_FL    CREDIT  \
15218      Состою в браке  ...               0                0  13450.00   
15219      Состою в браке  ...               0                0   8991.00   
15220  Не состоял в браке  ...               0                0   2700.00   
15221        Вдовец/Вдова  ...               0                0  24357.00   
15222      Состою в браке  ...               0                0  10245.00   

      TERM FACT_LIVING_TERM WORK_TIME  LOAN_NUM_PAYM LOAN_DLQ_NUM  \
15218   10              121     48.00             21            0   
15219    6              300    156.00              3            0   
15220    6               48     48.00              5            0   
15221   10              270       NaN              3            0   
15222    6              432     36.00              3            0   

      LOAN_MAX_DLQ_AMT          DATE_CREDIT  
15218             0.00  2018-04-11 17:38:21  
15219             0.00  2021-12-21 18:09:58  
15220             0.00  2016-11-26 10:13:32  
15221             0.00  2017-04-07 11:27:54  
15222             0.00  2018-03-12 18:26:55  

[5 rows x 37 columns]

Исходя из первичного просмотра было выявлено:

*   фрейм содержит 15223 записей, 37 колонок
*   во многих колонках недостает записей


*   уже понятно, что есть аномальные значения
*   какие типы данных содержатся




Пока данные не менялись, дальнейшая предобработка будет произведена в следующем этапе

**Шаг 2. Первичная обработка данных**

***Скорректировать заголовки***

*Смена названий заголовков*

In [ ]:
df.columns = new_columns = ['id','target','age','is_employed','is_pensioner','gender','childs','dependants','education','marital_status','industry',
               'job_title','org_type_state','org_type_capital','job_field','family_income','personal_income','reg_province','fact_address',
               'postal_province','tp_province','region','has_real_state','own_auto','rus_auto','has_house','has_cottage',"has_garage",'has_land',
               'credit','term','residence_duration','work_experience','number_of_payments','overdue_payments','max_of_amount','credit_date']

df.describe()

id    target       age  is_employed  is_pensioner    gender  \
count     15223.00  15223.00  15223.00     15223.00      15223.00  15223.00   
mean   65401830.70      0.12     40.41         0.91          0.13      0.65   
std     4568181.14      0.32     11.60         0.29          0.34      0.48   
min    59910150.00      0.00     21.00         0.00          0.00      0.00   
25%    61920915.00      0.00     30.00         1.00          0.00      0.00   
50%    64370996.00      0.00     39.00         1.00          0.00      1.00   
75%    67088017.50      0.00     50.00         1.00          0.00      1.00   
max    75292242.00      1.00     67.00         1.00          1.00      1.00   

         childs  dependants  personal_income  has_real_state  ...  \
count  15223.00    15223.00         15223.00        15223.00  ...   
mean       1.10        0.65         13853.84            0.31  ...   
std        1.00        0.81          9015.47            0.46  ...   
min        0.00        0.00            24.00            0.00  ...   
25%        0.00        0.00          8000.00            0.00  ...   
50%        1.00        0.00         12000.00            0.00  ...   
75%        2.00        1.00         17000.00            1.00  ...   
max       10.00        7.00        250000.00            1.00  ...   

       has_cottage  has_garage  has_land     credit      term  \
count     15223.00    15223.00  15223.00   15223.00  15223.00   
mean          0.01        0.02      0.04   14667.96      8.10   
std           0.10        0.14      0.20   12147.87      4.09   
min           0.00        0.00      0.00    2000.00      3.00   
25%           0.00        0.00      0.00    6500.00      6.00   
50%           0.00        0.00      0.00   11550.00      6.00   
75%           0.00        0.00      0.00   19170.00     10.00   
max           1.00        1.00      1.00  119700.00     36.00   

       residence_duration  work_experience  number_of_payments  \
count            15223.00         13855.00            15223.00   
mean              3039.34           292.21                7.88   
std             262455.44         24364.83                7.96   
min                -26.00             1.00                1.00   
25%                 41.00            24.00                4.00   
50%                108.00            48.00                5.00   
75%                204.00           110.00                8.00   
max           28101997.00       2867959.00              114.00   

       overdue_payments  max_of_amount  
count          15223.00       15223.00  
mean               0.20         243.02  
std                0.69         836.10  
min                0.00           0.00  
25%                0.00           0.00  
50%                0.00           0.00  
75%                0.00           0.00  
max               13.00       28012.20  

[8 rows x 23 columns]

*Настройка отображения чисел (убрать экспоненциальную запись) для более компактного понятного вида*

In [ ]:
pd.options.display.float_format = "{:.2f}".format
df.describe()

id   target      age  is_employed  is_pensioner   gender  \
count    15223.00 15223.00 15223.00     15223.00      15223.00 15223.00   
mean  65401830.70     0.12    40.41         0.91          0.13     0.65   
std    4568181.14     0.32    11.60         0.29          0.34     0.48   
min   59910150.00     0.00    21.00         0.00          0.00     0.00   
25%   61920915.00     0.00    30.00         1.00          0.00     0.00   
50%   64370996.00     0.00    39.00         1.00          0.00     1.00   
75%   67088017.50     0.00    50.00         1.00          0.00     1.00   
max   75292242.00     1.00    67.00         1.00          1.00     1.00   

        childs  dependants  personal_income  has_real_state  ...  has_cottage  \
count 15223.00    15223.00         15223.00        15223.00  ...     15223.00   
mean      1.10        0.65         13853.84            0.31  ...         0.01   
std       1.00        0.81          9015.47            0.46  ...         0.10   
min       0.00        0.00            24.00            0.00  ...         0.00   
25%       0.00        0.00          8000.00            0.00  ...         0.00   
50%       1.00        0.00         12000.00            0.00  ...         0.00   
75%       2.00        1.00         17000.00            1.00  ...         0.00   
max      10.00        7.00        250000.00            1.00  ...         1.00   

       has_garage  has_land    credit     term  residence_duration  \
count    15223.00  15223.00  15223.00 15223.00            15223.00   
mean         0.02      0.04  14667.96     8.10             3039.34   
std          0.14      0.20  12147.87     4.09           262455.44   
min          0.00      0.00   2000.00     3.00              -26.00   
25%          0.00      0.00   6500.00     6.00               41.00   
50%          0.00      0.00  11550.00     6.00              108.00   
75%          0.00      0.00  19170.00    10.00              204.00   
max          1.00      1.00 119700.00    36.00         28101997.00   

       work_experience  number_of_payments  overdue_payments  max_of_amount  
count         13855.00            15223.00          15223.00       15223.00  
mean            292.21                7.88              0.20         243.02  
std           24364.83                7.96              0.69         836.10  
min               1.00                1.00              0.00           0.00  
25%              24.00                4.00              0.00           0.00  
50%              48.00                5.00              0.00           0.00  
75%             110.00                8.00              0.00           0.00  
max         2867959.00              114.00             13.00       28012.20  

[8 rows x 23 columns]

*Изменение регистра, пробелов*

In [ ]:
df.columns = (
    df.columns
    .str.lower()
    .str.strip()
)

df.head()

id  target  age  is_employed  is_pensioner  gender  childs  \
0  59910150       0   49            1             0       1       2   
1  59910230       0   32            1             0       1       3   
2  59910525       0   52            1             0       1       4   
3  59910803       0   39            1             0       1       1   
4  59911781       0   30            1             0       0       0   

   dependants            education  marital_status  ... has_garage has_land  \
0           1  Среднее специальное  Состою в браке  ...          0        0   
1           3              Среднее  Состою в браке  ...          0        0   
2           0     Неполное среднее  Состою в браке  ...          0        0   
3           1               Высшее  Состою в браке  ...          0        0   
4           0              Среднее  Состою в браке  ...          0        0   

    credit term residence_duration work_experience  number_of_payments  \
0  8000.00    6                220           18.00                   6   
1 21650.00    6                137           97.00                   6   
2 33126.00   12                251           84.00                  11   
3  8491.82    6                 36          168.00                   6   
4 21990.00   12                 83          101.00                  16   

  overdue_payments max_of_amount          credit_date  
0                2       1580.00  2017-10-01 15:09:09  
1                1       4020.00  2020-10-09 18:58:35  
2                0          0.00  2015-05-05 11:10:07  
3                3       1590.00  2016-06-19 10:05:49  
4                2       2230.00  2016-03-13 14:22:06  

[5 rows x 37 columns]

*Дубликаты*

In [51]:
df.duplicated(subset="id").sum()
dups_all_except_id = df.duplicated(subset=[col for col in df.columns if col !="id"])



AttributeError: 'function' object has no attribute 'sum'

*Аномальные значения*